include/property.h

```diff
@@ -0,0 +1,181 @@
+    struct CodeContext {
+        std::string file;
+        std::string func;
+        int line;
+
+        CodeContext() = default;
+        CodeContext(const std::string& a, const std::string& b, const int& c): file(a), func(b), line(c) {}
+
+        decltype(auto) tie() const{
+            return std::tie(file, func, line);
+        }
+        bool operator<(const CodeContext& rhs) const {
+            return tie() < rhs.tie();
+        }
+    };
+    static const std::string CodeContextStart ("");
+    static const std::string CodeContextSep (" ");
+    static const std::string CodeContextEnd ("");
+    std::ostream& operator<<(std::ostream& s, const CodeContext& c) {
+        return s << CodeContextStart
+            << c.file << CodeContextSep
+            << c.func << CodeContextSep
+            << c.line << CodeContextEnd;
+    }
+
+    // TODO: RunContext
+
+    class LogRecord {
+    public:
+        using State = CodeContext;
+        LogRecord() : code() {};
+
+        friend std::ostream& operator<<(std::ostream& s, const LogRecord& c);
+
+        CodeContext code;
+        std::string message;
+
+        State get_state() {
+            return code;
+        }
+
+    private:
+        // TODO: make code private and try to make read_prop friendly
+        // NOTE: lmgtfy.com?q=one cannot declare a namespace within a class 
+        // RunContext prog;
+    };
+
+    // ### Methods to read Properties for the analyser
+
+    template<log_properties N>
+    void read_prop(LogRecord& p, const std::string& s);
+
+    template<>
+    void read_prop<FILE>(LogRecord& p, const std::string& s){
+        std::istringstream(s) >> p.code.file;
+    }
+    template<>
+    void read_prop<FUNC>(LogRecord& p, const std::string& s){
+        std::istringstream(s) >> p.code.func;
+    }
+
+    template<>
+    void read_prop<LINE>(LogRecord& p, const std::string& s){
+        std::istringstream(s) >> p.code.line;
+    }
+
+    template<>
+    void read_prop<MESG>(LogRecord& p, const std::string& s){
+        p.message = s;
+    }
+
+    // TODO: StringRef -> forward iterator on std::string
+    template<StringRef PS>
+    void read_props(LogRecord& p, PS) {
+    }
+
+    template<StringRef PS,log_properties head, log_properties... I>
+    void read_props(LogRecord& p, PS ps) {
+        read_prop<head>(p, *ps);
+        ps++;
+        read_props<PS,I...>(p, ps);
+    }
+
+    template<log_properties... I>
+    void parse_props(LogRecord& p, std::string& line) {
+        std::smatch m;
+        std::regex_match(line, m, logger::format_regex);
+        read_props<std::smatch::iterator, I...>(p, m.begin()+1);
+    }
+
+    static std::string LogRecordStart ("(");
+    static std::string LogRecordSep (" ");
+    static std::string LogRecordEnd (")");
+    std::ostream& operator<<(std::ostream& s, const LogRecord& c) {
+        return s << LogRecordStart
+            << c.code << LogRecordSep
+            << c.message << LogRecordEnd;
+    }
+
+
```

Comment:
`t2.cpp` is temporary code, showing how this should be used in logger.h.
I would suggest we keep it, since it won't be instantiated unless we use it. Final call is yours :)